File: agentictx/backend/app/schemas/discovery.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.discovery import (
    ActivityStatus,
    ClusterStatus,
    JTDStatus,
    MessageRole,
    RawInputType,
    ScopeStatus,
)


# ─── Raw Input ───────────────────────────────────────────────────────────────

class RawInputRead(BaseModel):
    id: uuid.UUID
    use_case_id: uuid.UUID
    type: RawInputType
    content: str | None
    file_path: str | None
    file_name: str | None
    mime_type: str | None
    processed: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# ─── Conversation Message ─────────────────────────────────────────────────────

class ConversationMessageRead(BaseModel):
    id: uuid.UUID
    use_case_id: uuid.UUID
    role: MessageRole
    content: Any  # list of Anthropic content blocks
    created_at: datetime

    model_config = {"from_attributes": True}


# ─── Activity (Job To Be Done) ──────────────────────────────────────────────

class ActivityRead(BaseModel):
    id: uuid.UUID
    use_case_id: uuid.UUID
    description: str
    system_context: str | None
    process_phase_id: uuid.UUID | None
    status: ActivityStatus
    linked_cognitive_jtd_id: uuid.UUID | None
    source_message_id: uuid.UUID | None = None
    is_modified: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# Keep old name as alias for backward compatibility
LivedJTDRead = ActivityRead


class ActivityCreate(BaseModel):
    description: str = Field(..., min_length=1)
    system_context: str | None = None
    process_phase_id: uuid.UUID | None = None


# Keep old name as alias for backward compatibility
LivedJTDCreate = ActivityCreate


class ActivityUpdate(BaseModel):
    description: str | None = Field(None, min_length=1)
    system_context: str | None = None
    process_phase_id: uuid.UUID | None = None
    status: ActivityStatus | None = None
    linked_cognitive_jtd_id: uuid.UUID | None = None
    is_modified: bool | None = None


# Keep old name as alias for backward compatibility
LivedJTDUpdate = ActivityUpdate


# ─── Cognitive Load ──────────────────────────────────────────────────────────

class CognitiveLoadRead(BaseModel):
    id: uuid.UUID
    use_case_id: uuid.UUID
    description: str
    cognitive_zone: str | None
    load_intensity: int | None
    process_phase_id: uuid.UUID | None
    linked_lived_jtd_ids: list[str] | None
    status: ActivityStatus
    source_message_id: uuid.UUID | None = None
    is_modified: bool = False
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# Keep old name as alias for backward compatibility
CognitiveJTDRead = CognitiveLoadRead


class CognitiveLoadCreate(BaseModel):
    description: str = Field(..., min_length=1)
    cognitive_zone: str | None = None
    load_intensity: int | None = Field(None, ge=0, le=3)
    process_phase_id: uuid.UUID | None = None


# Keep old name as alias for backward compatibility
CognitiveJTDCreate = CognitiveLoadCreate


class CognitiveLoadUpdate(BaseModel):
    description: str | None = Field(None, min_length=1)
    cognitive_zone: str | None = None
    load_intensity: int | None = Field(None, ge=0, le=3)
    process_phase_id: uuid.UUID | None = None
    linked_lived_jtd_ids: list[str] | None = None
    status: ActivityStatus | None = None
    is_modified: bool | None = None


# Keep old name as alias for backward compatibility
CognitiveJTDUpdate = CognitiveLoadUpdate


# ─── Agent Scope (Delegation Cluster) ────────────────────────────────────────

class AgentScopeRead(BaseModel):
    id: uuid.UUID
    use_case_id: uuid.UUID
    name: str
    purpose: str | None
    cognitive_jtd_ids: list[uuid.UUID]
    lived_jtd_ids: list[uuid.UUID]
    suitability_scores: dict[str, int] | None
    delegation_mode: str | None
    status: ScopeStatus
    is_scored: bool
    created_at: datetime
    updated_at: datetime


# Keep old name as alias for backward compatibility
DelegationClusterRead = AgentScopeRead


class AgentScopeUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    purpose: str | None = None
    delegation_mode: str | None = None
    status: ScopeStatus | None = None


# Keep old name as alias for backward compatibility
DelegationClusterUpdate = AgentScopeUpdate


# ─── Cognitive Map (full GET response) ───────────────────────────────────────

class CognitiveMapRead(BaseModel):
    use_case_id: uuid.UUID
    raw_inputs: list[RawInputRead]
    conversation_messages: list[ConversationMessageRead]
    lived_jtds: list[ActivityRead]
    cognitive_jtds: list[CognitiveLoadRead]
    delegation_clusters: list[AgentScopeRead]


# ─── Readiness Score ─────────────────────────────────────────────────────────

READINESS_DIMENSIONS = [
    "cognitive_load_intensity",
    "input_data_structure",
    "actionability_tool_coverage",
    "decision_determinism",
    "risk_compliance_sensitivity",
    "context_complexity",
    "exception_rate",
    "turn_taking_complexity",
    "latency_constraints",
]

# Keep old name as alias for backward compatibility
SUITABILITY_DIMENSIONS = READINESS_DIMENSIONS


class ReadinessScores(BaseModel):
    cognitive_load_intensity: int = Field(..., ge=0, le=3)
    input_data_structure: int = Field(..., ge=0, le=3)
    actionability_tool_coverage: int = Field(..., ge=0, le=3)
    decision_determinism: int = Field(..., ge=0, le=3)
    risk_compliance_sensitivity: int = Field(..., ge=0, le=3)
    context_complexity: int = Field(..., ge=0, le=3)
    exception_rate: int = Field(..., ge=0, le=3)
    turn_taking_complexity: int = Field(..., ge=0, le=3)
    latency_constraints: int = Field(..., ge=0, le=3)


# Keep old name as alias for backward compatibility
SuitabilityScores = ReadinessScores


# ─── Process Visualisation ────────────────────────────────────────────────────

class ProcessStepCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    sequence_order: int = Field(..., ge=0)
    is_breakpoint: bool = False
    cognitive_load_intensity: int | None = Field(None, ge=0, le=3)


class ProcessStepUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    sequence_order: int | None = Field(None, ge=0)
    is_breakpoint: bool | None = None
    cognitive_load_intensity: int | None = Field(None, ge=0, le=3)


class ProcessStepRead(BaseModel):
    id: uuid.UUID
    use_case_id: uuid.UUID
    name: str
    description: str | None
    sequence_order: int
    is_breakpoint: bool
    cognitive_load_intensity: int | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ScopeProcessStepRead(BaseModel):
    """Links an agent scope to a process step."""
    id: uuid.UUID
    cluster_id: uuid.UUID
    process_step_id: uuid.UUID

    model_config = {"from_attributes": True}


# Keep old name as alias for backward compatibility
ClusterProcessStepRead = ScopeProcessStepRead


class ProcessFlowRead(BaseModel):
    use_case_id: uuid.UUID
    steps: list[ProcessStepRead]
    cluster_steps: list[ScopeProcessStepRead]

File: agentictx/frontend/src/types/discovery.ts
// ─── Enums ──────────────────────────────────────────────────────────────────

export type RawInputType = "transcript" | "document" | "image" | "note";

export type ActivityStatus = "proposed" | "confirmed" | "rejected";

export type ScopeStatus = "proposed" | "confirmed" | "replaced";

export type MessageRole = "user" | "assistant" | "system";

// Backward compatibility aliases
export type JTDStatus = ActivityStatus;
export type ClusterStatus = ScopeStatus;

// ─── Source Material ─────────────────────────────────────────────────────────

export interface SourceMaterial {
  id: string;
  use_case_id: string;
  type: RawInputType;
  content: string | null;
  file_path: string | null;
  file_name: string | null;
  mime_type: string | null;
  processed: boolean;
  created_at: string;
}

// Backward compatibility alias
export type RawInput = SourceMaterial;

// ─── Conversation Message ─────────────────────────────────────────────────────

export interface ConversationMessage {
  id: string;
  use_case_id: string;
  role: MessageRole;
  content: AnthropicContentBlock[];
  created_at: string;
}

// Anthropic content block types (simplified subset used in UI)
export type AnthropicContentBlock =
  | { type: "text"; text: string }
  | { type: "tool_use"; id: string; name: string; input: Record<string, unknown> }
  | { type: "tool_result"; tool_use_id: string; content: unknown };

// ─── Activity (formerly LivedJTD) ───────────────────────────────────────────

export interface Activity {
  id: string;
  use_case_id: string;
  description: string;
  system_context: string | null;
  process_phase_id: string | null;
  status: ActivityStatus;
  linked_cognitive_jtd_id: string | null;
  source_message_id: string | null;
  is_modified: boolean;
  created_at: string;
  updated_at: string;
}

export interface ActivityCreate {
  description: string;
  system_context?: string | null;
  process_phase_id?: string | null;
}

export interface ActivityUpdate {
  description?: string;
  system_context?: string | null;
  process_phase_id?: string | null;
  status?: ActivityStatus;
  linked_cognitive_jtd_id?: string | null;
  is_modified?: boolean;
}

// Backward compatibility aliases
export type LivedJTD = Activity;
export type LivedJTDCreate = ActivityCreate;
export type LivedJTDUpdate = ActivityUpdate;

// ─── Cognitive Load (formerly CognitiveJTD) ─────────────────────────────────

export interface CognitiveLoad {
  id: string;
  use_case_id: string;
  description: string;
  cognitive_zone: string | null;
  load_intensity: number | null;
  process_phase_id: string | null;
  linked_lived_jtd_ids: string[] | null;
  status: ActivityStatus;
  source_message_id: string | null;
  is_modified: boolean;
  created_at: string;
  updated_at: string;
}

export interface CognitiveLoadCreate {
  description: string;
  cognitive_zone?: string | null;
  load_intensity?: number | null;
  process_phase_id?: string | null;
}

export interface CognitiveLoadUpdate {
  description?: string;
  cognitive_zone?: string | null;
  load_intensity?: number | null;
  process_phase_id?: string | null;
  linked_lived_jtd_ids?: string[] | null;
  status?: ActivityStatus;
  is_modified?: boolean;
}

// Backward compatibility aliases
export type CognitiveJTD = CognitiveLoad;
export type CognitiveJTDCreate = CognitiveLoadCreate;
export type CognitiveJTDUpdate = CognitiveLoadUpdate;

// ─── Agent Scope (formerly DelegationCluster) ───────────────────────────────

export interface ReadinessScores {
  cognitive_load_intensity: number;
  input_data_structure: number;
  actionability_tool_coverage: number;
  decision_determinism: number;
  risk_compliance_sensitivity: number;
  context_complexity: number;
  exception_rate: number;
  turn_taking_complexity: number;
  latency_constraints: number;
}

export interface AgentScope {
  id: string;
  use_case_id: string;
  name: string;
  purpose: string | null;
  cognitive_jtd_ids: string[];
  lived_jtd_ids: string[] | null;
  suitability_scores: ReadinessScores | null;
  delegation_mode: string | null;
  status: ScopeStatus;
  is_scored: boolean;
  created_at: string;
  updated_at: string;
}

export interface AgentScopeUpdate {
  name?: string;
  purpose?: string | null;
  cognitive_jtd_ids?: string[];
  lived_jtd_ids?: string[] | null;
  delegation_mode?: string | null;
  status?: ScopeStatus;
}

// Backward compatibility aliases
export type SuitabilityScores = ReadinessScores;
export type DelegationCluster = AgentScope;
export type DelegationClusterUpdate = AgentScopeUpdate;

// ─── Cognitive Map ────────────────────────────────────────────────────────────

export interface CognitiveMap {
  use_case_id: string;
  raw_inputs: SourceMaterial[];
  conversation_messages: ConversationMessage[];
  lived_jtds: Activity[];
  cognitive_jtds: CognitiveLoad[];
  delegation_clusters: AgentScope[];
}

// ─── WebSocket Event Types ────────────────────────────────────────────────────

export interface WSTextDelta {
  type: "text_delta";
  delta: string;
}

export interface WSActivitiesProposed {
  type: "activities_proposed";
  items: Activity[];
}

export interface WSCognitiveLoadProposed {
  type: "cognitive_load_proposed";
  items: CognitiveLoad[];
}

export interface WSAgentScopeProposed {
  type: "cluster_proposed";
  cluster: AgentScope;
}

export interface WSMessageComplete {
  type: "message_complete";
  message_id: string;
}

export interface WSError {
  type: "error";
  message: string;
}

export interface WSSystemNotification {
  type: "system_notification";
  text: string;
  highlight?: string;
}

export interface WSScopesReplaced {
  type: "clusters_replaced";
  count: number;
}

export interface WSToolCallStarted {
  type: "tool_call_started";
  tool_name: string;
}

export interface WSToolCallCompleted {
  type: "tool_call_completed";
  tool_name: string;
  summary: string;
}

export interface WSProcessPhasesProposed {
  type: "process_phases_proposed";
  phases: Phase[];
}

export type WSServerEvent =
  | WSTextDelta
  | WSActivitiesProposed
  | WSCognitiveLoadProposed
  | WSProcessPhasesProposed
  | WSAgentScopeProposed
  | WSMessageComplete
  | WSError
  | WSSystemNotification
  | WSScopesReplaced
  | WSToolCallStarted
  | WSToolCallCompleted;

// Backward compatibility aliases for WS events (kept for any external consumers)
export type WSLivedJTDsProposed = WSActivitiesProposed;
export type WSCognitiveJTDsProposed = WSCognitiveLoadProposed;
export type WSClusterProposed = WSAgentScopeProposed;
export type WSClustersReplaced = WSScopesReplaced;

// ─── Phase (formerly ProcessStep) ───────────────────────────────────────────

export interface Phase {
  id: string;
  use_case_id: string;
  name: string;
  description: string | null;
  sequence_order: number;
  is_breakpoint: boolean;
  cognitive_load_intensity: number | null;
  created_at: string;
  updated_at: string;
}

export interface PhaseCreate {
  name: string;
  description?: string | null;
  sequence_order: number;
  is_breakpoint?: boolean;
  cognitive_load_intensity?: number | null;
}

export interface PhaseUpdate {
  name?: string;
  description?: string | null;
  sequence_order?: number;
  is_breakpoint?: boolean;
  cognitive_load_intensity?: number | null;
}

// Backward compatibility aliases
export type ProcessStep = Phase;
export type ProcessStepCreate = PhaseCreate;
export type ProcessStepUpdate = PhaseUpdate;

export interface ScopePhaseLink {
  id: string;
  cluster_id: string;
  process_step_id: string;
}

// Backward compatibility alias
export type ClusterProcessStep = ScopePhaseLink;

export interface ProcessFlow {
  use_case_id: string;
  steps: Phase[];
  cluster_steps: ScopePhaseLink[];
}

// ─── UI-only types ─────────────────────────────────────────────────────────────

/** A chat message shown in the input panel. */
export interface ChatMessage {
  id: string;
  role: MessageRole;
  text: string;
  streaming?: boolean;
}
